### ISSUE TITLE
Cannot deploy on WSL 2 using Docker

### ISSUE THREAD

**Issue by jpmcb** [kubernetes-sigs/cluster-api#4891]
**What steps did you take and what happened:**
1. Install Windows Subsystem Linux on a compatible Windows machine including all requirements:
    - Windows 10 build: `19041.1052`
    - WSL Ubuntu version: `20.04.2 LTS (Focal Fossa)`
    - WSL kernal version: `Linux 5.4.72-microsoft-standard-WSL2 x86_64`
    - Docker desktop version (with WSL integration enabled): `20.10.7`
    - Clusterapi version: `0.3.20`
    - kubectl version: `1.21.1`
    - kind version: `0.11.1`
   
2. Following the quickstart guide, installed a kind cluster:
```
kind create cluster
```
3. Initiate a new bootstrap cluster
```
clusterctl init --infrastructure docker
```
4. Attempt to create a new worker cluster
```
clusterctl config cluster wsl-worker --flavor development --kubernetes-version v1.19.7 --control-plane-machine-count=1 --worker-machine-cou
nt=1 | kubectl apply -f -
```

At this point, the installation hangs and the cluster never exists the "provisioning" state. 

I notice the following log repeated in the `capd-manager`:
```
E0708 01:50:25.221097       1 controller.go:257] controller-runtime/controller "msg"="Reconciler error" "error"="failed to create helper for managing the externalLoadBalancer: failed to list containers: failed to list nodes: command \"docker ps -q -a --no-trunc --filter label=io.x-k8s.kind.cluster --format '{{.Names}}\\t{{.Label \"io.x-k8s.kind.cluster\"}}\\t{{.Image}}' --filter label=io.x-k8s.kind.cluster=wsl-worker --filter label=io.x-k8s.kind.role=external-load-balancer\" failed with error: exit status 1" "controller"="dockercluster" "name"="wsl-worker" "namespace"="default"
```
Also with these logs repeated in the `capi-manager`:
```
I0708 01:51:37.755644       1 machine_controller_noderef.go:45] controllers/Machine "msg"="Cannot reconcile Machine's Node, no valid ProviderID yet" "machine"="wsl-worker-md-0-f5448dbff-925jn" "namespace"="default"
```

The `machinedeployment` is stuck in the `ScalingUp` phase.  And everything else looks stuck. Not sure where else to look to help debug. Happy to dive into any of the logs as needed.

**What did you expect to happen:**
Is WSL expected to be supported? 

I know it's a bit of a bespoke linux distribution and can be challenging to work with since it sits atop windows networks, but users with Windows machines who want to develop on Linux without a full VM setup often choose this method. And _in theory_, it shouldn't be too different from a normal Linux / Docker setup.

/kind bug

**Comment by sbueringer** [comment]
@jpmcb I don't know if it's supported, but we can try in any case :)

I think you're on the right track with the log message from capd-manager. Unfortunately, CAPD doesn't seem to log any stdout/stderr. Two options to get more details:
* Exec into capd manager and try to execute the same command as in the error message. I'm hopeful there might be an explanation in stdout/stderr
* Take a look at the code of CAPD and add something which prints stdout/stderr

The first options is probably the more reasonable one for now.


Completely different idea. @stmcginnis is currently working on replacing the docker exec calls through Docker API calls, so it might be a good idea to try a CAPD manager based on his PR: https://github.com/kubernetes-sigs/cluster-api/pull/4413

**Comment by jpmcb** [comment]
@sbueringer I've tried to get a forward shell to the `manager` container, but I'm not sure that the docker container has any available shell:

```
 k exec -it \
  -n capd-system \
  -c manager capd-controller-manager-965f9fd59-pm68k -- /bin/bash
error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "901c93ed705c60c0020be0e86f64562aefc2d95b42b932705c6f17c140677ca2": OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "/bin/bash": stat /bin/bash: no such file or directory: unknown

k exec -it \
  -n capd-system \
  -c manager capd-controller-manager-965f9fd59-pm68k -- /bin/sh
error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "ea6bb5724be866807c56326ac4e72a609c2e5f1e0420842ba7c96d4006466736": OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory: unknown

k exec -it \
  -n capd-system \
  -c manager capd-controller-manager-965f9fd59-pm68k -- sh
error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "1cd06ea13d6602632af9ba1e1c7ef358d829218c4b5511c32c05131715a50bc9": OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "sh": executable file not found in $PATH: unknown
```
Any ideas how I could get a forward shell easily to the `manager` container?

**Comment by sbueringer** [comment]
@jpmcb 
I think the only alternative is to `docker exec` into the control plane node container and then `docker exec` again into `capd manager`, but not sure if that works and it's not really easier, just different.

I think the problem is that the capd manager is based on a distroless image. So I think you have to build your own capd-manager image based on e.g. debian. Docker file is located in: `test/infrastructure/docker/Dockerfile` , image can be build via: `make -C test/infrastructure/docker docker-build` (Disclaimer: I never tried to build the image with another base image myself)

**Comment by jpmcb** [comment]
Getting the bash forward shell to the control plane node works fine:
```
$ docker exec -it kind-control-plane bash
```
but once there, I don't see a way to exec onto any of the containers it knows about. There's doesn't seem to be a docker binary to use.

I can try to build using a different base image. Once I have a newly built image, what's the best way to inject it into the cluster-api cluster creation workflow?

**Comment by sbueringer** [comment]
@jpmcb Sorry my bad. Now I actually tried it. `crictl` is installed inside the kind image. You should be able to use `crictl exec`. I tried it with `crictl exec -it fa879fdd69622 /usr/bin/docker info`. The currently released v0.3.x container should only have the controller, `/usr/bin/kubectl` and `/usr/bin/docker`.

So I think you now have the following options to reproduce:

1. docker exec into the management cluster container + `crictl exec -it .... /usr/bin/docker` 

2. In case you want to exec into the capd-manager:
    1. build a custom image with another base image
    2. edit the `capd-controller-manager` deployment in `capd-system` after `clusterctl init` and replace the image
    3. Either `kubectl exec` or `docker exec` + `crictl exec` into the `capd-controller-manager`

**Comment by jpmcb** [comment]
I haven't found alot of time to debug this and actually don't have my windows environment anymore.

/unassign

Is there anyone from the community (maybe a go to windows person) who could look at this?
